presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java

```diff
@@ -0,0 +1,517 @@
+            Arrays.fill(hashTables, -1);
+
+            verify(positionCount < offsets.length, "incorrect offsets size");
+
+            for (int i = 0; i < positionCount; i++) {
+                int keyOffset = offsets[i];
+                int keyCount = offsets[i + 1] - keyOffset;
+                if (keyCount < 0) {
+                    throw new IllegalArgumentException(format("Offset is not monotonically ascending. offsets[%s]=%s, offsets[%s]=%s", i, offsets[i], i + 1, offsets[i + 1]));
+                }
+                if (mapIsNull != null && mapIsNull[i] && keyCount != 0) {
+                    throw new IllegalArgumentException("A null map must have zero entries");
+                }
+                buildHashTable(
+                        keyBlock,
+                        keyOffset,
+                        keyCount,
+                        keyBlockHashCode,
+                        hashTables,
+                        keyOffset * HASH_MULTIPLIER,
+                        keyCount * HASH_MULTIPLIER);
+            }
+            set(hashTables);
+        }
+    }
+
+    @Override
+    public Block getBlockUnchecked(int internalPosition)
+    {
+        assert internalPositionInRange(internalPosition, this.getOffsetBase(), getPositionCount());
+
+        int startEntryOffset = getOffsets()[internalPosition];
+        int endEntryOffset = getOffsets()[internalPosition + 1];
+        return new SingleMapBlock(internalPosition - getOffsetBase(), startEntryOffset * 2, (endEntryOffset - startEntryOffset) * 2, this);
+    }
+
+    @Override
+    public boolean isNullUnchecked(int internalPosition)
+    {
+        assert mayHaveNull() : "no nulls present";
+        assert internalPositionInRange(internalPosition, this.getOffsetBase(), getPositionCount());
+        return getMapIsNull()[internalPosition];
+    }
+
+    @Override
+    public Block appendNull()
+    {
+        boolean[] mapIsNull = appendNullToIsNullArray(getMapIsNull(), getOffsetBase(), getPositionCount());
+        int[] offsets = appendNullToOffsetsArray(getOffsets(), getOffsetBase(), getPositionCount());
+
+        return createMapBlockInternal(
+                getOffsetBase(),
+                getPositionCount() + 1,
+                Optional.of(mapIsNull),
+                offsets,
+                getRawKeyBlock(),
+                getRawValueBlock(),
+                getHashTables());
+    }
+
+    private void calculateLogicalSize()
+    {
+        int entriesStart = getOffset(0);
+        int entriesEnd = getOffset(getPositionCount());
+        int entryCount = entriesEnd - entriesStart;
+        logicalSizeInBytes = getRawKeyBlock().getRegionLogicalSizeInBytes(entriesStart, entryCount) +
+                getRawValueBlock().getRegionLogicalSizeInBytes(entriesStart, entryCount) +
+                (Integer.BYTES + Byte.BYTES) * (long) this.getPositionCount() +
+                Integer.BYTES * HASH_MULTIPLIER * (long) entryCount;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).